builder/builder.go

```diff
@@ -522,137 +522,135 @@ func (b *Builder) BuildChroots(template *x509.Certificate, privkey *rsa.PrivateK
 	}
 
-	err = ioutil.WriteFile(b.Statedir+"/image/LAST_VER", []byte(b.Mixver), 0644)
+	//
+	// Create update metadata for the mix.
+	//
+	updatecmd := exec.Command(prefixflag+"swupd_create_update", "-S", b.Statedir, "--minversion", strconv.Itoa(minVersion), "-F", b.Format, "--osversion", b.Mixver)
+	updatecmd.Stdout = os.Stdout
+	updatecmd.Stderr = os.Stderr
+	err = updatecmd.Run()
 	if err != nil {
-		helpers.PrintError(err)
-		os.Exit(1)
+		return errors.Wrapf(err, "failed to create update metadata")
 	}
 
-	if b.Upstreamurl != "" {
-		fmt.Println("Saving the upstream version URL " + b.Upstreamurl)
-		upstreamurl := b.Statedir + "/www/" + b.Mixver + "/upstreamurl"
-		err = ioutil.WriteFile(upstreamurl, []byte(b.Upstreamurl), 0644)
+	// Clean up the bundle chroots as only the full chroot is needed from this point on.
+	if !keepChroots {
+		var files []os.FileInfo
+		files, err = ioutil.ReadDir(b.Bundledir)
 		if err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
+			fmt.Fprintf(os.Stderr, "Ignored error when cleaning bundle chroots: %s", err)
+		}
+		basedir := filepath.Join(b.Statedir, "image", b.Mixver)
+		for _, f := range files {
+			if f.Name() == "full" {
+				continue
+			}
+			err = os.RemoveAll(filepath.Join(basedir, f.Name()))
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Ignored error when cleaning bundle chroots: %s", err)
+			}
 		}
-	}
-	fmt.Println("Saving the upstream version " + b.Clearver)
-	upstreamver := b.Statedir + "/www/" + b.Mixver + "/upstreamver"
-	err = ioutil.WriteFile(upstreamver, []byte(b.Clearver), 0644)
-	if err != nil {
-		helpers.PrintError(err)
-		os.Exit(1)
 	}
 
-}
-
-// CleanChroots will remove chroots based on what bundles are defined
-func (b *Builder) CleanChroots() {
-	files := helpers.GetDirContents(b.Bundledir)
-	basedir := b.Statedir + "/image/" + b.Mixver + "/"
-
-	for _, f := range files {
-		if f.Name() == "full" {
-			continue
-		}
-		err := os.RemoveAll(basedir + f.Name())
+	//
+	// Sign the Manifest.MoM that was just created.
+	//
+	if !skipSigning {
+		err = b.SignManifestMoM()
 		if err != nil {
-			helpers.PrintError(err)
+			return err
 		}
+		fmt.Println("Signed Manifest.MoM")
 	}
-}
 
-// BuildUpdate will produce an update consumable by the swupd client
-func (b *Builder) BuildUpdate(prefixflag string, minvflag int, formatflag string, signflag bool, publishflag bool, keepchrootsflag bool) error {
-	if formatflag != "" {
-		b.Format = formatflag
+	//
+	// Create full files.
+	//
+	fullfilecmd := exec.Command(prefixflag+"swupd_make_fullfiles", "-S", b.Statedir, b.Mixver)
+	fullfilecmd.Stdout = os.Stdout
+	fullfilecmd.Stderr = os.Stderr
+	err = fullfilecmd.Run()
+	if err != nil {
+		return errors.Wrapf(err, "couldn't create fullfiles")
 	}
 
-	if _, err := os.Stat(b.Statedir + "/www/version/format" + b.Format); os.IsNotExist(err) {
-		if err = os.MkdirAll(b.Statedir+"/www/version/format"+b.Format, 0777); err != nil {
-			helpers.PrintError(err)
-			return err
-		}
+	//
```

Comment:
por que

Reply:
Idea was to separate the main sections, but it failed to accomplish that. Removed.